pkg/storage/chunk/prometheus_chunk.go

```diff
@@ -162,24 +218,57 @@ func (p *prometheusChunkIterator) Value() model.SamplePair {
 	}
 }
 
-// TODO native histograms support, assumes it's used on float chunk and only keeps floats
-// The Batch function takes a valueType since the first sample we process is already acquired
-// by a Next/Seek/etc and we cannot get its type otherwise.
+func (p *prometheusChunkIterator) AtHistogram() (int64, *histogram.Histogram) {
+	return p.it.AtHistogram()
+}
+
+func (p *prometheusChunkIterator) AtFloatHistogram() (int64, *histogram.FloatHistogram) {
+	return p.it.AtFloatHistogram()
+}
+
+func (p *prometheusChunkIterator) Timestamp() int64 {
+	return p.it.AtT()
+}
+
 func (p *prometheusChunkIterator) Batch(size int, valueType chunkenc.ValueType) Batch {
 	var batch Batch
-	j := 0
-	for j < size {
-		if valueType == chunkenc.ValFloat {
+	batch.ValueType = valueType
+	var populate func(j int)
+	switch valueType {
+	case chunkenc.ValFloat:
+		populate = func(j int) {
 			t, v := p.it.At()
 			batch.Timestamps[j] = t
 			batch.Values[j] = v
-			j++
 		}
+	case chunkenc.ValHistogram:
+		populate = func(j int) {
+			t, h := p.it.AtHistogram()
+			batch.Timestamps[j] = t
+			batch.PointerValues[j] = unsafe.Pointer(h)
+		}
+	case chunkenc.ValFloatHistogram:
+		populate = func(j int) {
+			t, fh := p.it.AtFloatHistogram()
+			batch.Timestamps[j] = t
+			batch.PointerValues[j] = unsafe.Pointer(fh)
+		}
+	default:
+		panic(fmt.Sprintf("invalid chunk encoding %v", valueType))
```

Comment:
I can add a return error instead here , marginal impact on performance:
```
cpu: AMD Ryzen 7 4700U with Radeon Graphics         
                                                                                                   │ benchmark-batch-pointers.txt │     benchmark-batch-error.txt      │
                                                                                                   │            sec/op            │   sec/op     vs base               │
NewChunkMergeIterator_CreateAndIterate/chunks:_1000_samples_per_chunk:_100_duplication_factor:_1-8                    4.731m ± 1%   4.856m ± 1%  +2.63% (p=0.000 n=10)
NewChunkMergeIterator_CreateAndIterate/chunks:_1000_samples_per_chunk:_100_duplication_factor:_3-8                    14.35m ± 1%   14.47m ± 2%       ~ (p=0.165 n=10)
NewChunkMergeIterator_CreateAndIterate/chunks:_100_samples_per_chunk:_100_duplication_factor:_1-8                     484.3µ ± 1%   500.6µ ± 1%  +3.35% (p=0.000 n=10)
NewChunkMergeIterator_CreateAndIterate/chunks:_100_samples_per_chunk:_100_duplication_factor:_3-8                     1.455m ± 1%   1.471m ± 1%  +1.16% (p=0.023 n=10)
NewChunkMergeIterator_CreateAndIterate/chunks:_1_samples_per_chunk:_100_duplication_factor:_1-8                       12.24µ ± 3%   12.30µ ± 3%       ~ (p=0.592 n=10)
NewChunkMergeIterator_CreateAndIterate/chunks:_1_samples_per_chunk:_100_duplication_factor:_3-8                       32.18µ ± 2%   32.74µ ± 2%  +1.72% (p=0.015 n=10)
geomean                                                                                                               515.8µ        524.6µ       +1.70%

                                                                                                   │ benchmark-batch-pointers.txt │      benchmark-batch-error.txt      │
                                                                                                   │             B/op             │     B/op      vs base               │
NewChunkMergeIterator_CreateAndIterate/chunks:_1000_samples_per_chunk:_100_duplication_factor:_1-8                   73.43Ki ± 0%   73.46Ki ± 0%  +0.04% (p=0.000 n=10)
NewChunkMergeIterator_CreateAndIterate/chunks:_1000_samples_per_chunk:_100_duplication_factor:_3-8                   282.9Ki ± 0%   283.0Ki ± 0%  +0.03% (p=0.000 n=10)
NewChunkMergeIterator_CreateAndIterate/chunks:_100_samples_per_chunk:_100_duplication_factor:_1-8                    9.586Ki ± 0%   9.617Ki ± 0%  +0.33% (p=0.000 n=10)
NewChunkMergeIterator_CreateAndIterate/chunks:_100_samples_per_chunk:_100_duplication_factor:_3-8                    34.24Ki ± 0%   34.34Ki ± 0%  +0.27% (p=0.000 n=10)
NewChunkMergeIterator_CreateAndIterate/chunks:_1_samples_per_chunk:_100_duplication_factor:_1-8                      2.062Ki ± 0%   2.094Ki ± 0%  +1.52% (p=0.000 n=10)
NewChunkMergeIterator_CreateAndIterate/chunks:_1_samples_per_chunk:_100_duplication_factor:_3-8                      4.859Ki ± 0%   4.953Ki ± 0%  +1.93% (p=0.000 n=10)
geomean                                                                                                              20.22Ki        20.36Ki       +0.68%

                                                                                                   │ benchmark-batch-pointers.txt │      benchmark-batch-error.txt       │
                                                                                                   │          allocs/op           │  allocs/op   vs base                 │
NewChunkMergeIterator_CreateAndIterate/chunks:_1000_samples_per_chunk:_100_duplication_factor:_1-8                    1.013k ± 0%   1.013k ± 0%       ~ (p=1.000 n=10) ¹
NewChunkMergeIterator_CreateAndIterate/chunks:_1000_samples_per_chunk:_100_duplication_factor:_3-8                    3.023k ± 0%   3.023k ± 0%       ~ (p=1.000 n=10) ¹
NewChunkMergeIterator_CreateAndIterate/chunks:_100_samples_per_chunk:_100_duplication_factor:_1-8                      113.0 ± 0%    113.0 ± 0%       ~ (p=1.000 n=10) ¹
NewChunkMergeIterator_CreateAndIterate/chunks:_100_samples_per_chunk:_100_duplication_factor:_3-8                      323.0 ± 0%    323.0 ± 0%       ~ (p=1.000 n=10) ¹
NewChunkMergeIterator_CreateAndIterate/chunks:_1_samples_per_chunk:_100_duplication_factor:_1-8                        14.00 ± 0%    14.00 ± 0%       ~ (p=1.000 n=10) ¹
NewChunkMergeIterator_CreateAndIterate/chunks:_1_samples_per_chunk:_100_duplication_factor:_3-8                        26.00 ± 0%    26.00 ± 0%       ~ (p=1.000 n=10) ¹
geomean                                                                                                                185.5         185.5       +0.00%
```

Reply:
On the other hand, we're testing if we've covered all enumeration values. Which should be done by a linter I think.